### PR TITLE
Duplication of content and ingest items

### PR DESCRIPTION
**Pull request by marwoodandrew**
[SD-1290] As a User, I want the ability to fetch previously fetched content.
[SD-815] As a user, I want to duplicate content items

Major changes 
* Fetch action allowed from both ingest and archive.
* Fetch is allowed more than once from ingest.
* New guids are created with each ingest.
* Original guids are saved in new field ingest_id
* New field family_id allows related items to be identified.
* All versions are duplicated along with content items (deep copy)

### CODE REVIEW

File: apps/archive/archive.py [hunk 1]
```diff
@@ -36,6 +37,8 @@
 from copy import copy
 import superdesk
 import logging
+from apps.common.models.utils import get_model
+from ..item_lock.models.item import ItemModel
```
Comment (amagdas) [inline apps/archive/archive.py]: Better use absolute imports.

File: apps/archive/archive.py [hunk 2]
```diff
@@ -82,9 +85,7 @@ class ArchiveVersionsResource(Resource):
 
 
 class ArchiveVersionsService(BaseService):
-    def on_create(self, docs):
-        for doc in docs:
-            doc['versioncreated'] = utcnow()
+    pass
```
Comment (amagdas) [inline apps/archive/archive.py]: Is this class still needed?

File: apps/archive/archive.py [hunk 3]
```diff
@@ -265,6 +266,46 @@ def restore_version(self, id, doc):
...
+        item_model = get_model(ItemModel)
+        on_duplicate_item(new_doc)
+        item_model.create([new_doc])
+        self.duplicate_versions(original_doc['guid'], new_doc)
+        if new_doc.get('state') != 'submitted':
+            get_resource_service('tasks').patch(new_doc['_id'], {'state': 'submitted'})
+        return new_doc['guid']
+
+    def duplicate_versions(self, old_id, new_doc):
+        lookup = {}
```
Comment (amagdas) [inline apps/archive/archive.py]: `{'guid': old_id}`

File: apps/archive/archive.py [hunk 4]
```diff
@@ -265,6 +266,46 @@ def restore_version(self, id, doc):
...
+        old_versions = get_resource_service('archive_versions').get(req=None, lookup=lookup)
+
+        for old_version in old_versions:
+            old_version['_id_document'] = new_doc['_id']
+            del old_version['_id']
+            old_version['guid'] = new_doc['guid']
+            old_version['unique_name'] = new_doc['unique_name']
+            old_version['unique_id'] = new_doc['unique_id']
+            old_version['versioncreated'] = utcnow()
+            get_resource_service('archive_versions').post([old_version])
```
Comment (amagdas) [inline apps/archive/archive.py]: We can use buld insert here, `.post(old_versions)` should be more efficient, especially if there are many versions.

**Comment by amagdas** [comment]
Please use `git rebase` on feature branches as much as possible so we end up reviewing only the new code.

File: apps/archive/archive.py (hunk 1 above)
Comment (akintolga) [inline apps/archive/archive.py]: Done.

File: apps/archive/archive.py (hunk 2 above)
Comment (akintolga) [inline apps/archive/archive.py]: I wanted to keep it because the end_point "archive_versions" likely to be needed again.

Comment (sivakuna-aap) [inline apps/archive/archive.py]: We already have endpoint for fetching archive versions. We don't need a new endpoint for `archive_versions`.

File: apps/archive/common.py [hunk 5]
```diff
@@ -44,10 +44,23 @@ def on_create_item(docs):
         if 'unique_id' not in doc:
             generate_unique_id_and_name(doc)
 
+        if 'family_id' not in doc:
+            doc['family_id'] = doc['guid']
+
```
Comment (sivakuna-aap) [inline apps/archive/common.py]: It can be `doc.setdefault('family_id', doc['guid'])`

File: apps/archive/archive.py (hunk 3 above)
Comment (akintolga) [inline apps/archive/archive.py]: Done.

File: apps/archive/archive.py (hunk 4 above)
Comment (akintolga) [inline apps/archive/archive.py]: Agreed. Thanks.

**Pull request closed without merging**